### PR TITLE
Kobo: Initial Elipsa 2E support

### LINKED ISSUE #1048: Koreader does not show up in possible apps on Android
it would be great if Koreader shows up as possible opening app for ebook formats on android. That is not needed on einks (non android ones). Right now the only way to open a book is through the KR file browser.

thanks

### LINKED ISSUE #10408: KOREader does not start on Kobo Elipsa 2E
* KOReader version: v2023.04
* Device: Kobo Elipsa 2E

#### Issue
KOReader can't start on Kobo Elipsa 2E.
I can see the following error in the log file: `./luajit: frontend/device/kobo/device.lua:1560: unrecognized Kobo model condor with device id 389`

#### Steps to reproduce
Install KOReader on a Kobo Elipsa 2E device, with firmware `4.36.21095` and try to start it.
The device will freeze for 10+ seconds and then it will reboot.

This is the relevant crash log portion (full file attached)
<details>
  <summary>crash.log</summary>

```
Original fb rotation is set @ 1
Original fb bitdepth is set @ 32bpp
Switching fb bitdepth to 8bpp & rotation to Portrait
[FBInk] Unidentified Kobo device code (389)!
[FBInk] Detected a Kobo Unknown! (389 => ?? @ Mark ?)
[FBInk] Clock tick frequency appears to be 100 Hz
[FBInk] Screen density set to 167 dpi
[FBInk] Variable fb info: 1404x1872, 32bpp @ rotation: 1 (Clockwise, 90°)
[FBInk] Fixed fb info: ID is "hwtcon", length of fb mem: 10543104 bytes & line length: 5616 bytes
[FBInk] Canonical rotation: 2 (Upside Down, 180°)
[FBInk] Fontsize set to 16x16 (IBM (Default) base glyph size: 8x8)
[FBInk] Line length: 87 cols, Page size: 117 rows
[FBInk] Detected a change in framebuffer rotation (1 -> 3)
[FBInk] Reinitializing...
[FBInk] Variable fb info: 1404x1872, 32bpp @ rotation: 3 (Counter Clockwise, 270°)
[FBInk] Fixed fb info: ID is "hwtcon", length of fb mem: 10543104 bytes & line length: 5616 bytes
[FBInk] Canonical rotation: 0 (Upright, 0°)
[FBInk] Fontsize set to 16x16 (IBM (Default) base glyph size: 8x8)
[FBInk] Line length: 87 cols, Page size: 117 rows
[FBDepth] Screen is 1404x1872 (1404x1877 addressable, fb says 1404x1872)
[FBDepth] Buffer is mapped for 10543104 bytes with a scanline stride of 5616 bytes
[FBDepth] Requested canonical rota 0 translates to 3 for this device
[FBDepth] Switching fb to 8bpp @ rotation 3 . . .
Updating bitdepth from 32bpp to 8bpp
Sanitizing grayscale flag for 8bpp
Updating rotation from 1 (Clockwise, 90°) to 3 (Counter Clockwise, 270°)
Bitdepth is now 32bpp (grayscale: 1) @ rotate: 3 (Counter Clockwise, 270°)
[FBDepth] Screen is 1404x1872 (1404x1877 addressable, fb says 1404x1872)
[FBDepth] Buffer is mapped for 10543104 bytes with a scanline stride of 5616 bytes
---------------------------------------------
                launching...
  _  _____  ____                _
 | |/ / _ \|  _ \ ___  __ _  __| | ___ _ __
 | ' / | | | |_) / _ \/ _` |/ _` |/ _ \ '__|
 | . \ |_| |  _ <  __/ (_| | (_| |  __/ |
 |_|\_\___/|_| \_\___|\__,_|\__,_|\___|_|

 It's a scroll... It's a codex... It's KOReader!

 [*] Current time: 05/09/23-14:27:19
 [*] Version: v2023.04

ffi.load: libs/libutf8proc.so.2
ffi.load: blitbuffer
./luajit: frontend/device/kobo/device.lua:1560: unrecognized Kobo model condor with device id 389
stack traceback:
	[C]: in function 'error'
	frontend/device/kobo/device.lua:1560: in main chunk
	[C]: in function 'probeDevice'
	frontend/device.lua:61: in main chunk
	[C]: in function 'require'
	./reader.lua:151: in main chunk
	[C]: at 0x000140bd
!!!!
Uh oh, something went awry... (Crash n°1: 05/09/23 @ 14:27:19)
Running FW 4.36.21095 on Linux 4.9.77 (#1 SMP PREEMPT 39a784229346-20230331T162906-B0331164118)
Attempting to restart KOReader . . .
!!!!
```
</details>

It tries several times, and then it quits and crashes the entire system:
<details>
  <summary>crash.log</summary>

```
Too many consecutive crashes, aborting . . .
!!!! ! !!!!
Restoring original fb bitdepth @ 32bpp & rotation @ 1
[FBInk] Unidentified Kobo device code (389)!
[FBInk] Detected a Kobo Unknown! (389 => ?? @ Mark ?)
[FBInk] Clock tick frequency appears to be 100 Hz
[FBInk] Screen density set to 167 dpi
[FBInk] Variable fb info: 1404x1872, 32bpp @ rotation: 3 (Counter Clockwise, 270°)
[FBInk] Fixed fb info: ID is "hwtcon", length of fb mem: 10543104 bytes & line length: 5616 bytes
[FBInk] Canonical rotation: 0 (Upright, 0°)
[FBInk] Fontsize set to 16x16 (IBM (Default) base glyph size: 8x8)
[FBInk] Line length: 87 cols, Page size: 117 rows
[FBInk] Detected a change in framebuffer rotation (3 -> 1)
[FBInk] Reinitializing...
[FBInk] Variable fb info: 1404x1872, 32bpp @ rotation: 1 (Clockwise, 90°)
[FBInk] Fixed fb info: ID is "hwtcon", length of fb mem: 10543104 bytes & line length: 5616 bytes
[FBInk] Canonical rotation: 2 (Upside Down, 180°)
[FBInk] Fontsize set to 16x16 (IBM (Default) base glyph size: 8x8)
[FBInk] Line length: 87 cols, Page size: 117 rows
[FBDepth] Screen is 1404x1872 (1404x1877 addressable, fb says 1404x1872)
[FBDepth] Buffer is mapped for 10543104 bytes with a scanline stride of 5616 bytes
[FBDepth] Current bitdepth is already 32bpp, but the grayscale flag is bogus!
[FBDepth] Switching fb to 32bpp @ rotation 1 . . .
Updating bitdepth from 32bpp to 32bpp
Sanitizing grayscale flag for bitdepths > 8bpp
Updating rotation from 3 (Counter Clockwise, 270°) to 1 (Clockwise, 90°)
Bitdepth is now 32bpp (grayscale: 0) @ rotate: 1 (Clockwise, 90°)
[FBDepth] Screen is 1404x1872 (1404x1877 addressable, fb says 1404x1872)
[FBDepth] Buffer is mapped for 10543104 bytes with a scanline stride of 5616 bytes
```
</details>

I am available to test new device support, but I don't know how to add it myself.

[crash.log](https://github.com/koreader/koreader/files/11431066/crash.log)

### PR DESCRIPTION
**Pull request by NiLuJe**
Depends on https://github.com/koreader/koreader-base/pull/1642, Fix #10408

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/10719)

### CODE REVIEW

**Comment by NiLuJe** [comment]
For potential testers, I've just updated the [OCP thread](https://www.mobileread.com/forums/showthread.php?t=314220) with a KFMon build that should handle the basics (but may fail after an USBMS session, more on that later).

And since this requires a few built binaries, here's a delta package you can drop over a current nightly that should contain everything in this PR.
[KOReader-Elipsa2E.tar.gz](https://github.com/koreader/koreader/files/12131423/KOReader-Elipsa2E.tar.gz)

**Comment by NiLuJe** [comment]
Re: that KFMon issue, I would also very much like to take a look at a KFMon log, to see how the framebuffer evolves during the boot process.

So, ideally, install KFMon, try a simple USBMS session *in nickel*, eject, and send me the log (`/usr/local/kfmon/kfmon.log`).

----

Speaking of USBMS, the *KOreader* USBMS tool *should* also be working, but obviously I haven't tested that ;).

**Comment by ealex** [comment]
hello
koreader still crashes, but with a different error:
[crash.log](https://github.com/koreader/koreader/files/12157264/crash.log)


"USBMS session in nickel" = ? just a normal usb files transfer ?
this is the log:
[kfmon.log](https://github.com/koreader/koreader/files/12157267/kfmon.log)

**Comment by NiLuJe** [comment]
Looks like I fumbled the ffi decls, I'll take a look when I get back on Friday, thanks!

At least the KFMon stuff went better than I expected ;p.

**Comment by ealex** [comment]
> Looks like I fumbled the ffi decls, I'll take a look when I get back on Friday, thanks!
> 
> At least the KFMon stuff went better than I expected ;p.

i'll be away until 7 August

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 1 of 5 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/10719)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @NiLuJe)

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/10719)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @NiLuJe)

File: platform/kobo/enable-wifi.sh [hunk 1]
```diff
@@ -14,43 +14,84 @@ for fd in /proc/"$$"/fd/*; do
...
+        echo "0xDB9DB9" >/proc/driver/wmt_dbg
+        # Disable the LPBK test
+        echo "7 9 0" >/proc/driver/wmt_dbg
+        # Nickel appears to sleep for ~1s
+        sleep 1
+        echo "0xDB9DB9" >/proc/driver/wmt_dbg
+        # Enable the LPBK test (this'll block for ~1.3s)
+        echo "7 9 1" >/proc/driver/wmt_dbg
+        # Finally, power on the chip
+        echo 1 >/dev/wmtWifi
```
Comment (Frenzie) [inline platform/kobo/enable-wifi.sh:70]: Heh wtf

File: platform/kobo/enable-wifi.sh [hunk 2]
```diff
@@ -14,43 +14,84 @@ for fd in /proc/"$$"/fd/*; do
...
-fi
+        ;;
+esac
 # Moar sleep!
 usleep 250000
 
 # Load WiFi modules
 # NOTE: Used to be exported in WIFI_MODULE_PATH before FW 4.23
-if ! grep -q "^${WIFI_MODULE}" "/proc/modules"; then
+if ! grep -q "^${WIFI_MODULE} " "/proc/modules"; then
```
Comment (Frenzie) [inline platform/kobo/enable-wifi.sh:94]: Was it at risk of actually matching something it shouldn't or just for sanity?

**Review by Frenzie — APPROVED** [review]

File: platform/kobo/enable-wifi.sh (hunk 1 above)
Comment (NiLuJe) [reply platform/kobo/enable-wifi.sh:70]: Yeah, don't ask me what LPBK stands for ;D.

File: platform/kobo/enable-wifi.sh (hunk 2 above)
Comment (NiLuJe) [reply platform/kobo/enable-wifi.sh:94]: Currently, no, but a common prefix used for a series of interdependent modules is not terribly far fetched (and it was nearly the case here, which prompted that change ^^).

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 1 of 2 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/10719)* status: all files reviewed, 2 unresolved discussions (waiting on @Frenzie, @pazos, and @poire-z)

**Comment by NiLuJe** [comment]
Okay, going to merge this as I have something that needs wider testing based on top of that; and it can't be worse than nothing ;p.

The only thing that's unimplemented is *eject* detection in the USBMS tool (I'll need someone to check logread over WiFi for that one); but *unplug* detection should work. (Assuming the whole thing actually works, period, of course ;p).

That, and confirming how exactly Nickel powers off the device, but that's niche and mostly harmless, so it'll probably wait until I get access to a device on the same SoC myself.

**Comment by Knut0730** [comment]
Sorry for my ignorance, is it now possible to test it in the "easy way" by drag and drop onto my Kobo?

**Comment by Frenzie** [comment]
If you grab a build from https://build.koreader.rocks/download/nightly/

**Comment by Knut0730** [comment]
And thats perfectly safe for my device?

**Comment by NiLuJe** [comment]
We have yet to kill any device, ever ;o).

**Comment by pazos** [comment]
>> And thats perfectly safe for my device?

> We have yet to kill any device, ever ;o).

Also, see license

> 15. Disclaimer of Warranty.
THERE IS NO WARRANTY FOR THE PROGRAM, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES PROVIDE THE PROGRAM "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE PROGRAM IS WITH YOU. SHOULD THE PROGRAM PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL NECESSARY SERVICING, REPAIR OR CORRECTION.

> 16. Limitation of Liability.
IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MODIFIES AND/OR CONVEYS THE PROGRAM AS PERMITTED ABOVE, BE LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE PROGRAM (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A FAILURE OF THE PROGRAM TO OPERATE WITH ANY OTHER PROGRAMS), EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.

**Comment by ealex** [comment]
hello,

@Knut0730: the install won't break the device, but some things seem wrong.

first test: open the user manual (default start-up page) and try accessing the menus.
"bottom menu": seems to display both of them ( on the bottom of the screen:  rotation, page size, font, contrast, front-light) and the top menu. any press on it's icons activates the corresponding icon on the top menu.
"top menu": displays only the bottom menu.

basically all actions / buttons are mirrored on the screen - for example, the "X" mark to close the bookmarks screen is displayed on top-right, but I need to click the top-down corner for it to register.

this happens in all orientations - is there a specific file that defines the screen orientation ? i can test on my device and make a branch for the fix

**Comment by NiLuJe** [comment]
> is there a specific file that defines the screen orientation

It's a tad more complex than that ;).

Check what happens when drawing via `./finger_trace` as found in the following DevCap bundle, *after* running `./fbdepth -R UR`. (And, as usual, logo on the bottom right).

I'll need the full console log from that, as well as what's actually drawn on screen and where if it doesn't track your input ;).

**Comment by ealex** [comment]
hello
```
[admin@condor root]# ./fbdepth -R UR
[FBInk] Detected a Kobo Elipsa 2E (389 => Condor @ Mark 11)
[FBInk] This device does not support HW inversion
[FBInk] Enabled MediaTek quirks
[FBInk] Clock tick frequency appears to be 100 Hz
[FBInk] Screen density set to 227 dpi
[FBInk] Variable fb info: 1404x1872, 32bpp @ rotation: 1 (Clockwise, 90°)
[FBInk] Fixed fb info: ID is "hwtcon", length of fb mem: 10543104 bytes & line length: 5616 bytes
[FBInk] Canonical rotation: 0 (Upright, 0°)
[FBInk] Fontsize set to 24x24 (IBM (Default) base glyph size: 8x8)
[FBInk] Line length: 58 cols, Page size: 78 rows
[FBDepth] Screen is 1404x1872 (1404x1877 addressable, fb says 1404x1872)
[FBDepth] Buffer is mapped for 10543104 bytes with a scanline stride of 5616 bytes
[FBDepth] Requested canonical rota 0 translates to 1 for this device
[FBDepth] Current rotation is already 1!
```

then:
```
[admin@condor root]# ./finger_trace 
[FBInk] Detected a Kobo Elipsa 2E (389 => Condor @ Mark 11)
[FBInk] This device does not support HW inversion
[FBInk] Enabled MediaTek quirks
[FBInk] Clock tick frequency appears to be 100 Hz
[FBInk] Screen density set to 227 dpi
[FBInk] Variable fb info: 1404x1872, 32bpp @ rotation: 1 (Clockwise, 90°)
[FBInk] Fixed fb info: ID is "hwtcon", length of fb mem: 10543104 bytes & line length: 5616 bytes
[FBInk] Canonical rotation: 0 (Upright, 0°)
[FBInk] Fontsize set to 24x24 (IBM (Default) base glyph size: 8x8)
[FBInk] Line length: 58 cols, Page size: 78 rows
[FBInk] Pen colors set to #000000 for the foreground and #000000 for the background
Initialized FBInk v1.25.0-95-g27816e1 for Kobo
Initialized libevdev for device `Elan Touchscreen`
Rotation: 1 -> 0


top left: draw appears on bottom left
1691729167.001603 [ FINGER ] [DOWN] @ (1812,   49) -> (  49, 1812) => (  49, 1812)
libevdev error in sanitize_event: BUG: Device "Elan Touchscreen" received a double tracking ID 211 in slot 0.
1691729167.013294 [ FINGER ] [DOWN] @ Similar coordinates
libevdev error in sanitize_event: BUG: Device "Elan Touchscreen" received a double tracking ID 211 in slot 0.
1691729167.026080 [ FINGER ] [DOWN] @ Similar coordinates
libevdev error in sanitize_event: BUG: Device "Elan Touchscreen" received a double tracking ID 211 in slot 0.
1691729167.038903 [ FINGER ] [DOWN] @ Similar coordinates
libevdev error in sanitize_event: BUG: Device "Elan Touchscreen" received a double tracking ID 211 in slot 0.
1691729167.051420 [ FINGER ] [DOWN] @ Similar coordinates
libevdev error in sanitize_event: BUG: Device "Elan Touchscreen" received a double tracking ID 211 in slot 0.
1691729167.064067 [ FINGER ] [DOWN] @ Similar coordinates
libevdev error in sanitize_event: BUG: Device "Elan Touchscreen" received a double tracking ID 211 in slot 0.
1691729167.076708 [ FINGER ] [DOWN] @ Similar coordinates
libevdev error in sanitize_event: BUG: Device "Elan Touchscreen" received a double tracking ID 211 in slot 0.
1691729167.089322 [ FINGER ] [DOWN] @ Similar coordinates
libevdev error in sanitize_event: BUG: Device "Elan Touchscreen" received a double tracking ID 211 in slot 0.
1691729167.102014 [ FINGER ] [DOWN] @ Similar coordinates
libevdev error in sanitize_event: BUG: Device "Elan Touchscreen" received a double tracking ID 211 in slot 0.
1691729167.114679 [ FINGER ] [DOWN] @ Similar coordinates
libevdev error in sanitize_event: BUG: Device "Elan Touchscreen" received a double tracking ID 211 in slot 0.
...
```

It looks like everything is mirrored around this line:
mirror_line

top left => bottom left
bottom left => top left
top right => bottom right
bottom right => top right

top and bottom are mirrored around the center of the screen, left and right is preserved

**Comment by NiLuJe** [comment]
I love NTX kernels. >_<".

Does this one behave any better?

[Kobo-DevCap-Test.tar.gz](https://github.com/koreader/koreader/files/12320250/Kobo-DevCap-Test.tar.gz)

**Comment by ealex** [comment]
> I love NTX kernels. >_<".
> 
> Does this one behave any better?
> 
> [Kobo-DevCap-Test.tar.gz](https://github.com/koreader/koreader/files/12320267/Kobo-DevCap-Test.tar.gz)

this version works

**Comment by NiLuJe** [comment]
Whee \o/

The following should make KOReader behave, then:

```diff
diff --git a/frontend/device/input.lua b/frontend/device/input.lua
index c316ca1eb..8f3b6a8ae 100644
--- a/frontend/device/input.lua
+++ b/frontend/device/input.lua
@@ -392,6 +392,20 @@ function Input:adjustABS_SwitchAxesAndMirrorX(ev, max_x)
     end
 end
 
+function Input:adjustABS_SwitchAxesAndMirrorY(ev, max_y)
+    if ev.code == C.ABS_X then
+        ev.code = C.ABS_Y
+        ev.value = max_y - ev.value
+    elseif ev.code == C.ABS_Y then
+        ev.code = C.ABS_X
+    elseif ev.code == C.ABS_MT_POSITION_X then
+        ev.code = C.ABS_MT_POSITION_Y
+        ev.value = max_y - ev.value
+    elseif ev.code == C.ABS_MT_POSITION_Y then
+        ev.code = C.ABS_MT_POSITION_X
+    end
+end
+
 function Input:adjustABS_Translate(ev, by)
     if ev.code == C.ABS_X or ev.code == C.ABS_MT_POSITION_X then
         ev.value = by.x + ev.value
diff --git a/frontend/device/kobo/device.lua b/frontend/device/kobo/device.lua
index d9d2a0676..8ea2c6e61 100644
--- a/frontend/device/kobo/device.lua
+++ b/frontend/device/kobo/device.lua
@@ -108,6 +108,8 @@ local Kobo = Generic:extend{
     touch_switch_xy = true,
     -- most Kobos have also mirrored X coordinates
     touch_mirrored_x = true,
+    -- but a few mirror on the Y axis instead
+    touch_mirrored_y = false,
     -- enforce portrait mode on Kobos
     --- @note: In practice, the check that is used for in ffi/framebuffer is no longer relevant,
     ---        since, in almost every case, we enforce a hardware Portrait rotation via fbdepth on startup by default ;).
@@ -516,6 +518,7 @@ local KoboCondor = Kobo:extend{
     display_dpi = 227,
     pressure_event = C.ABS_MT_PRESSURE,
     touch_mirrored_x = false,
+    touch_mirrored_y = true,
     hasNaturalLight = yes,
     frontlight_settings = {
         frontlight_white = "/sys/class/backlight/mxc_msp430.0/brightness",
@@ -975,7 +978,7 @@ end
 function Kobo:initEventAdjustHooks()
     -- Build a single composite hook, to avoid duplicated branches...
     local koboInputMangling
-    -- NOTE: touch_switch_xy is *always* true, but not touch_mirrored_x...
+    -- NOTE: touch_switch_xy is *always* true, but not touch_mirrored_x or touch_mirrored_y...
     if self.touch_switch_xy and self.touch_mirrored_x then
         local max_x = self.screen:getWidth() - 1
         koboInputMangling = function(this, ev)
@@ -985,7 +988,16 @@ function Kobo:initEventAdjustHooks()
                 gyroTranslation(ev)
             end
         end
-    elseif self.touch_switch_xy and not self.touch_mirrored_x then
+    elseif self.touch_switch_xy and self.touch_mirrored_y then
+        local max_y = self.screen:getHeight() - 1
+        koboInputMangling = function(this, ev)
+            if ev.type == C.EV_ABS then
+                this:adjustABS_SwitchAxesAndMirrorY(ev, max_y)
+            elseif ev.type == C.EV_MSC and ev.code == C.MSC_RAW then
+                gyroTranslation(ev)
+            end
+        end
+    elseif self.touch_switch_xy and not self.touch_mirrored_x and not self.touch_mirrored_y then
         koboInputMangling = function(this, ev)
             if ev.type == C.EV_ABS then
                 this:adjustABS_SwitchXY(ev)
```

**Comment by Knut0730** [comment]
> If you grab a build from https://build.koreader.rocks/download/nightly/

How do I install a nightly version on my device after downloading it from this site? Following the usual installation manual from the Github page doesn't work, tells me that it finds no files...

**Comment by mergen3107** [comment]
@Knut0730 
Switch channel to developer.

**Comment by NiLuJe** [comment]
@mergen3107: Not possible, the stable doesn't *boot* on those devices yet ;).

@Knut0730: You can't use the OCP scripts, you'll have to install manually.

**Comment by Knut0730** [comment]
> @mergen3107: Not possible, the stable doesn't *boot* on those devices yet ;).
> 
> @Knut0730: You can't use the OCP scripts, you'll have to install manually.

Why does it have to be so complicated heh, but it looks to be like most things are working and that a stable release is right around the corner?

**Comment by Frenzie** [comment]
Of course without testing who knows how stable it'll be. ;-P

But anyway, you can just use the script to automate the install of 2023.6 and overwrite the files in /mnt/sd/.adds/koreader with those from the downloaded file?

**Comment by Knut0730** [comment]
@Frenzie where do I find these files? I have now installed the 2023.6 version but cannot find the directory you speak of.

**Comment by Frenzie** [comment]
When you plug in the device that probably just means `.adds` in what's mounted. On Windows you should just see that, on Linux it's typically Ctrl + H to show hidden folders and I think that's something like Command + Shift + . on Mac.

**Comment by Knut0730** [comment]
Aha! Now I see! So I was supposed to do that BEFORE installing the 2023.6 version, what do I then do if I made the mistake of installing the build completely without replacing the files with the nightly build? Can I simply run the install file once again and it will just overwrite it? I also don't understand how I would go fourth with uninstalling KOReader.

**Comment by Frenzie** [comment]
No, after, otherwise there wouldn't be any files to overwrite and it wouldn't be set up with NickelMenu and all that. ^_^

The difficult part is setting up the device so it can start the program, which is automated by the script. But most of "installing" just means "extract these files to .adds/koreader".

**Comment by Knut0730** [comment]
Okay, now it works! KOReader lauches on my Elipsa 2E! Looks good! Now - anything in particular you need testing @NiLuJe ?

**Comment by Knut0730** [comment]
I found that when turning on Night Mode all ink goes away and it becomes impossible to navigate. Please fix 😅

**Comment by Knut0730** [comment]
Was #10841 to fix the Night Mode? After updating to the Aug 26 nightly build my WiFi connection works, but Night Mode still makes the screen blank.

**Comment by Knut0730** [comment]
I can also mention that when connecting a device (my Mac) to my Kobo I get the "_Could not start the USBMS session! The device will shut down in 90 sec._" message. This should probably also get fixed?

**Comment by NiLuJe** [comment]
> Was #10841 to fix the Night Mode? After updating to the Aug 26 nightly build my WiFi connection works, but Night Mode still makes the screen blank.

Try again after a full device restart for good measure.

If that doesn't help that would possibly point to wonky eclipse waveform modes, so I'll need a Nickel strace in Dark mode.



> I can also mention that when connecting a device (my Mac) to my Kobo I get the "_Could not start the USBMS session! The device will shut down in 90 sec._" message.

That's slightly earlier than I was expecting it to fail, but, ditto, I need syslogs from an USBMS session to deal with that.

**Comment by Knut0730** [comment]
> Try again after a full device restart for good measure.
I just did, a reboot did not help.

> If that doesn't help that would possibly point to wonky eclipse waveform modes, so I'll need a Nickel strace in Dark mode.
Anything I can help with? If yes, how?

> That's slightly earlier than I was expecting it to fail, but, ditto, I need syslogs from an USBMS session to deal with that.
(Since it'll trip a shutdown, you'll need to do that live, over Wi-Fi).

I cannot help much with getting the USBMS syslogs as I do not know how I get these. But I can attempt it if you give me some sort of instructions.

**Comment by NiLuJe** [comment]
You'd need SSH access, pretty much like what was described in the original FR and/or the linked Clara 2E thread.

**Comment by Knut0730** [comment]
Welp - hightling in Reflow isn't working as it should...

**Comment by NiLuJe** [comment]
>  Welp - hightling in Reflow isn't working as it should...

That shouldn't be device-specific, so I've moved it to a dedicated issue (#10854).

Please attach a file that reproduces the crash and detailed reproduction steps in there.

**Comment by NiLuJe** [comment]
> Was #10841 to fix the Night Mode? After updating to the Aug 26 nightly build my WiFi connection works, but Night Mode still makes the screen blank.

A nickel strace was provided via the MR Discord, and AFAICT, this *should* work (in fact, that revert is irrelevant, the MDP invert works).

So new, idea: I'd really like an actual look at how the screen looks like, and I'd also like you to try to do this from inside a book, and, if you toggle NM via the menu, to blindly *close* the menu, and try to page around the book normally.

And the matching verbose debug logs, in case there's something wonky happening and we're not actually uring the right constants for some reason ;o).
